persistence/nosql/persistence/metastore/src/main/java/org/apache/polaris/persistence/nosql/metastore/mutation/MutationAttempt.java

```diff
@@ -0,0 +1,593 @@
+              yield false;
+            }
+            // If 1, drop the last catalog role.
+            // Should be the catalog admin role, but don't validate this.
+            // (No need to drop the catalog role here, it'll be eventually done by
+            // persistence-maintenance!)
+
+            yield true;
+          }
+          case POLICY ->
+              memoizedIndexedAccess
+                  .referenceHead(POLICY_MAPPINGS_REF_NAME, PolicyMappingsObj.class)
+                  .map(
+                      policyMappingsObj -> {
+                        var index =
+                            policyMappingsObj
+                                .policyMappings()
+                                .indexForRead(persistence, POLICY_MAPPING_SERIALIZER);
+
+                        var prefixKey = policyIndexPrefixKey((PolicyObj) originalObj, entity);
+
+                        var iter = index.iterator(prefixKey, prefixKey, false);
+
+                        if (iter.hasNext() && !update.cleanup()) {
+                          mutationResults.dropResult(POLICY_HAS_MAPPINGS);
+                          return false;
+                        }
+
+                        while (iter.hasNext()) {
+                          var elem = iter.next();
+                          var key = PolicyMappingsObj.PolicyMappingKey.fromIndexKey(elem.getKey());
+                          var reversed = key.reverse();
+
+                          mutationResults.addPolicyIndexKeyToRemove(elem.getKey());
+                          mutationResults.addPolicyIndexKeyToRemove(reversed.toIndexKey());
+                        }
+
+                        return true;
+                      })
+                  .orElse(true);
+          default -> true;
+        };
+    if (ok) {
+      byId.remove(entityIdKey);
+      byName.remove(requireNonNull(originalNameKey));
+      mutationResults.dropResult(entity);
+
+      if (changes != null) {
+        changes.put(originalNameKey, ChangeRemove.builder().build());
+      }
+    }
+  }
+
+  private static void applyEntityUpdateMutation(
+      UpdateKeyForCatalogAndEntityType updateKeyForCatalogAndEntityType,
+      CommitterState<? extends ContainerObj, MutationResults> state,
+      UpdatableIndex<ObjRef> byName,
+      UpdatableIndex<IndexKey> byId,
+      UpdatableIndex<Change> changes,
+      UpdatableIndex<EntityIdSet> locations,
+      EntityUpdate update,
+      MutationResults mutationResults) {
+    var entity = update.entity();
+    var entityType = entity.getType();
+    var persistence = state.persistence();
+    var now = persistence.currentInstant();
+
+    var entityIdKey = IndexKey.key(entity.getId());
+    var originalNameKey = byId.get(entityIdKey);
+
+    var entityParentId = entity.getParentId();
+
+    if (originalNameKey == null) {
+      mutationResults.entityResult(ENTITY_NOT_FOUND);
+      return;
+    }
+    var originalRef = byName.get(originalNameKey);
```

Comment:
What are the scenarios that can lead to this `byName` not having this `originalNameKey` that we just got from `byId`? Would it indicate `byId` and `byName` getting out of sync somehow?

Reply:
There's no thinkable scenario actually.